### PR TITLE
Add "container" setting, to allow attaching AOS to any scrollable element

### PR DESCRIPTION
**Pull request by m-kutnik**
## Related Issue
Issue: #223

## Your solution
- Added `container` to AOS options (default: window). 
- Variable `container` is pointing to `HTML Element`, or to `Window` if querySelector returns null.
- EventListener is passing `container` to `handleScroll()` as a second argument
- `handleScroll()` evaluates scroll distance by using `Element.scrollTop` or `Window.pageYOffset`

### Defining scrollable element
```js
AOS.init({
  container: ".scrollable-div"
})
```
## How Has This Been Tested?
- By running included tests

In own projects, tested on:
- Desktop: Firefox 62, Chrome 71, Edge 42
- Android: Chrome, Firefox

### CODE REVIEW

**Comment by Lapsus** [comment]
i would really love to see that feature merged.

File: src/js/aos.js [hunk 1]
```diff
@@ -48,16 +49,21 @@ const isBrowserNotSupported = () => document.all && !window.atob;
 const initializeScroll = function initializeScroll() {
   // Extend elements objects in $aosElements with their positions
   $aosElements = prepare($aosElements, options);
+  // Define container element
+  let container =
```
Comment (ncoden) [inline src/js/aos.js]: It's not redefined later. Use `const` instead.

File: src/js/aos.js [hunk 2]
```diff
@@ -48,16 +49,21 @@ const isBrowserNotSupported = () => document.all && !window.atob;
...
+    options.container == 'window'
+      ? window
+      : document.querySelector(options.container) || window;
```
Comment (ncoden) [inline src/js/aos.js]: I don't think it's a great idea to only take `String` to target an element. Since this is executed synchronously, we should at least support a `HTMLElement`, if not only it.

Please use `window` (instead of `'window'`) for the default of `container`, and only search for the element if `options.container` is a string.

Comment (ncoden) [reply src/js/aos.js]: Also: if we have no results with `document.querySelector(options.container)`, there is something wrong because we cannot find the element the user want to target with the provided `container` option. We should throw an error in this case.

File: src/js/helpers/handleScroll.js [hunk 3]
```diff
@@ -78,9 +78,12 @@ const applyClasses = (el, top) => {
  * Scroll logic - add or remove 'aos-animate' class on scroll
  *
  * @param  {array} $elements         array of elements nodes
+ * @param  {node}  container         container
```
Comment (ncoden) [inline src/js/helpers/handleScroll.js]: The type is: `HTMLElement`.

See https://stackoverflow.com/a/21489264/4317384

File: src/js/helpers/handleScroll.js [hunk 4]
```diff
@@ -78,9 +78,12 @@ const applyClasses = (el, top) => {
...
  * @return {void}
  */
-const handleScroll = $elements =>
-  $elements.forEach((el, i) => applyClasses(el, window.pageYOffset));
+const handleScroll = ($elements, container) => {
+  let top = container == window ? container.pageYOffset : container.scrollTop;
```
Comment (ncoden) [inline src/js/helpers/handleScroll.js]: Please explain. Add a comment 😉

Comment (ncoden) [reply src/js/helpers/handleScroll.js]: `top` is not redefined later. Use `const`.

**Review by ncoden — CHANGES_REQUESTED** [review]

**Comment by ncoden** [comment]
@michalsnik Reviewed it for you.

File: src/js/helpers/handleScroll.js (hunk 3 above)
Comment (m-kutnik) [reply src/js/helpers/handleScroll.js]: `container` could be a `Window` too, so maybe `{(HTMLElement|Window)}`? 😃

File: src/js/aos.js (hunk 2 above)
Comment (m-kutnik) [reply src/js/aos.js]: Take a look and tell me what to add/change/fix/rename

```js
// Define container element
let container = isElementValid(options.container);
if (!container) throw `[AOS] Defined property 'container' is not valid`;

(...)

const isElementValid = function isElementValid(container) {
  if (container instanceof Element || container == window) return container;
  if (typeof container === 'string') {
    const queryResult = document.querySelector(container);
    if (queryResult) return queryResult;
  }
  return false;
};
```

And thank you sooo much for your time and help :)

Comment (m-kutnik) [reply src/js/aos.js]: and, of course, `let` -> `const` ;)

File: src/js/helpers/offsetCalculator.js [hunk 5]
```diff
@@ -10,9 +10,18 @@
...
 import getInlineOption from './getInlineOption';
+import validContainer from './container';
 
-export const getPositionIn = (el, defaultOffset, defaultAnchorPlacement) => {
-  const windowHeight = window.innerHeight;
+export const getPositionIn = (
+  el,
+  defaultOffset,
+  defaultAnchorPlacement,
+  container
```
Comment (michalsnik) [inline src/js/helpers/offsetCalculator.js]: I think the container should be the second argument here. It’s more important than offset or offsetAnchorPlacement

File: src/js/helpers/offsetCalculator.js [hunk 6]
```diff
@@ -10,9 +10,18 @@
...
+) => {
+  const windowHeight =
+    container === window
+      ? validContainer(container).innerHeight
+      : validContainer(container).clientHeight;
```
Comment (michalsnik) [inline src/js/helpers/offsetCalculator.js]: What do you think about creating “getContainer()” function in container.js that validates the container and returns an object with top offset and height, being a universal wrapper that allows us to not care about clientHeight or innerHeight here? We could then just use getContainer(container).topOffset

Or alternatively: getOffset(container) and getHeight(container)

File: src/js/aos.js [hunk 7]
```diff
@@ -16,6 +16,7 @@ import detect from './helpers/detector';
 import handleScroll from './helpers/handleScroll';
 import prepare from './helpers/prepare';
 import elements from './helpers/elements';
+import validContainer from './helpers/container';
```
Comment (michalsnik) [inline src/js/aos.js]: This function should probably be called “validateContainer” as this form describes the action being takes. On the other hand this not only validates the container but also returns the valid one. Maybe “getContainer” that internally uses “validateContainer” would be a better fit. What do you think?

**Review by michalsnik — CHANGES_REQUESTED** [review]
Hey! Thank you so much on working on this @m-kutnik :muscle: And thank you @ncoden for the help, I really appreciate it!

I have few suggestions as well as requests before we proceed with this feature. Could you please add a relevant E2E test that would make sure the whole functionality works as expected alongside other features? If you’ll have troubles with this let me know - I can do it myself too :)
And the second thing - could you please document this feature in readme?

I have few edge cases in mind and will try to pull the branch and test it locally too.

File: src/js/helpers/handleScroll.js (hunk 3 above)
Comment (ncoden) [reply src/js/helpers/handleScroll.js]: Yes https://stackoverflow.com/a/35351628/4317384

File: src/js/aos.js (hunk 2 above)
Comment (ncoden) [reply src/js/aos.js]: The error message must give the user entry, tell what is wrong and what is expected.
```diff
- [AOS] Defined property 'container' is not valid
+ AOS - cannot find the container element "xxx". The container option must be an HTMLElement or a CSS Selector.
```

This function doesn't returns a Boolean but performs a search. Use `find...` instead of `is...`.
```diff
- let container = isElementValid(options.container);
+ let container = findElement(options.container);
```

This function returns the element if it is found. Returns nothing if nothing is found.
```diff
- return false;
+ return null;
```

Also, add some documentation on top of this helper :)

File: src/js/aos.js (hunk 7 above)
Comment (ncoden) [reply src/js/aos.js]: See my comment at https://github.com/michalsnik/aos/pull/393#discussion_r222465362

Also, I think it would be better in this case to keep error throwing at the root, instead of nested in an helper.

File: src/js/helpers/offsetCalculator.js (hunk 6 above)
Comment (ncoden) [reply src/js/helpers/offsetCalculator.js]: This seems like overengineering for where a small refactor would be enough.

- Why is there argument validation in an helper ? Shouldn't helpers be generic 
? This seems required because `getPositionIn` is called with `options.container` because the `prepare` helper only receive `options`. We should have the container validated and resolved in the main function and then passed in to all helpers.

- If we want some abstraction to deal with `innerHeight` and `clientHeight`, this have nothing to do with `container` resolution. I don't think that's really useful, but the best way to do it would be a generic `getElementHeight` helper.

Comment (ncoden) [reply src/js/helpers/offsetCalculator.js]: > We should have the container validated and resolved in the main function and then passed in to all helpers.

If the helpers API looks weird with `container` in, then they are not generic helpers and there should be regrouped with the main function in a single `AOS` class.

An other solution (much easier but not really cleaner) would be to replace `option.container` with the resolved container.

Comment (m-kutnik) [reply src/js/helpers/offsetCalculator.js]: Great idea, will do

Comment (michalsnik) [reply src/js/helpers/offsetCalculator.js]: I agree that passing resolved container + using generic helper for getting height would be the cleanest solution here. Thanks for an additional insight @ncoden 

I don't see a problem with having a generic helper that takes `Element` as an argument.

File: src/js/aos.js (hunk 7 above)
Comment (michalsnik) [reply src/js/aos.js]: The reason I proposed `getContainer` is that the current function is not void, and in fact it returns resolved container - which might be a bit misleading.
Now that I think about it once again, I think we can better rename it to `resolveContainer` as it seems more natural - and then below if there is no `container` we can throw an Error at the root.

**Comment by m-kutnik** [comment]
Updated and ready for code review @michalsnik @ncoden 👍

**Comment by miduraa11** [comment]
Hello
Is this feature will be avaible soon? Really waiting for it :)

**Comment by xavibonell** [comment]
Why this never got merged? This is a 'must-have' feature, otherwise, the library gets pretty much unusable in a lot of scenarios.

**Comment by m-kutnik** [comment]
@xavibonell who knows :D 
I removed my fork of the repo, but I can find files on my disk if needed. Right now I don't have enough free time to commit to this repo :/ (but maybe in future)

Oh and sorry @michalsnik for re-review request I sent today- misclick :D

**Comment by ashishkumar468** [comment]
I see that this feature is not merged yet, it would be great to have the ability to specify the container, seems like @m-kutnik  is away if it is okay, I am willing to resolve the merge conflicts and any other code-review comments. It would be great if someone could confirm that if it is feasible to merge the PR.

**Review by eliortabeka — APPROVED** [review]

**Review by wvffle — APPROVED** [review]

**Comment by kennarddh** [comment]
Will this be released?

**Review by wesley-pd — APPROVED** [review]

**Comment by Staijn1** [comment]
Would also love to see this merged, any update if it is feasible to merge?
Happy to contribute.

**Comment by mpkelly** [comment]
Folks, you can just switch to Sal which works in a similar way. It doesn't have the shortcoming of only working with the window. Here is how I use in NextJS

```jsx
'use client'

import React, { useEffect } from 'react'
import sal from 'sal.js'
import 'sal.js/dist/sal.css'

export const SALInit = ({ selector }) => {
  const [root, setRoot] = React.useState()

  useEffect(() => {
    if (!root) {
      return
    }

    sal({
      root,
    })

  }, [root])

  return (
    <div
      style={{ display: 'none' }}
      ref={(ref) => {
        if (ref) {
          setRoot(document.querySelector(selector))
        }
      }}
    />
  )
}

```

You can then annotate your elements in the root container like so 

```jsx
  <div
    className={'flex gap-4 rounded-md bg-orange-50 p-4'}
    data-sal="zoom-in"
    data-sal-delay="500"
  >
```
